File: Algorithms/Search/1_BinarySearch/C/binarySearch.c
#include <stdio.h>
#include <stdlib.h>

int binarySearch(int array[], int left, int right, int x);

int main() {

    int array[] = {2, 3, 4, 10, 40};
    int n = sizeof(array) / sizeof(array[0]);
    int x = 10;
    int result = binarySearch(array, 0, n - 1, x) - 1;
    (result == -1) ? printf("Elemento não encontrado.\n") : printf("Elemento encontrado na posição %d.\n", result);
    return 0;
}

 int binarySearch(int array[], int left, int right, int x){

    if (right >= left) {
        int mid = left + (right - left) / 2;

        if (array[mid] == x) {
            return mid;
        }

        if (array[mid] > x) {
            return binarySearch(array, left, mid - 1, x);
        }

        return binarySearch(array, mid + 1, right, x);
    }

    return -1;
}

File: Algorithms/Search/1_BinarySearch/Rust/binary_search.rs
fn binary_search(arr: Vec<i32>, target: i32) -> i32 {
    let mut left = 0; 
    let mut right = arr.len() as i32 - 1;
    while left <= right {
        let mid = left + (right - left) / 2;
        if arr[mid as usize] == target {
            return mid;
        } else if arr[mid as usize] < target {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    -1
}

fn main() {
    let arr = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    let target = 5;
    let result = binary_search(arr, target);
    println!("Index of target: {}", result);
}
 
